### ISSUE TITLE
Unable to build library using grunt. Fatal error: Protocol error (Runtime.callFunctionOn): Object reference chain is too long.

### ISSUE THREAD

**Issue by malviys** [processing/p5.js#5367]
#### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [x] Build tools and processes
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Friendly error system
- [ ] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

#### Details about the bug:
Unable to build library using grunt

- p5.js version: 1.4.0
- Web browser and version: Google Chrome Version 92.0.4515.107 (Official Build) (64-bit)
- Operating System: Windows 10 Build 19043.1110
- Node version: 16.5.0 
- Steps to reproduce this:
```bash
npm run grunt
```

Console error:
```bash
Fatal error: Protocol error (Runtime.callFunctionOn): Object reference chain is too long
```

**Comment by limzykenneth** [comment]
@malviys Can you share a more complete console output?

**Comment by malviys** [comment]
@limzykenneth It is very long output Grunt builds and runs test all test cases. Don't know how much code would be sufficient because it tries to run test cases and then fails.

**Comment by limzykenneth** [comment]
Basically we need to know which task failed and if it's the test, which test is causing the issue. I know it can be difficult to tell where the issue occur at first glance so if you could share a more complete output I can diagnose it better. You can put it on a [gist](https://gist.github.com/) or pastebin as those are a bit more suitable for long text than github issue comments.

**Comment by malviys** [comment]
@limzykenneth here is the pastebin link https://pastebin.com/FdULWrF2 for the error details

**Comment by limzykenneth** [comment]
It seems to be a puppeteer problem. You can change your node.js version to 14.x and npm version to 6.x and you should be able to run `grunt` fully. I'll need to see what kind of fix is available for npm version 7.x.

**Comment by limzykenneth** [comment]
I couldn't do a straightforward update to puppeteer 10.2.0 (latest version) because of a few failed unit tests that I'm not too familiar with. @outofambit Can you perhaps have a look?

```
  1) Global Error Handling
       identifies TypeError 'readFromNull':

      AssertionError: expected 0 to equal 1
      + expected - actual

      -0
      +1
      
      at Function.assert.strictEqual (http://localhost:9001/node_modules/chai/chai.js:2329:32)
      at http://localhost:9001/test/unit/core/error_helpers.js:774:14

  2) Global Error Handling
       identifies TypeError 'readFromUndefined':

      AssertionError: expected 0 to equal 1
      + expected - actual

      -0
      +1
      
      at Function.assert.strictEqual (http://localhost:9001/node_modules/chai/chai.js:2329:32)
      at http://localhost:9001/test/unit/core/error_helpers.js:792:14

  3) time and date
       p5.prototype.millis
         result should be greater than running time:
     AssertionError: everything is ok: expected 50 to be > 50
```

**Comment by limzykenneth** [comment]
The unit tests are now fixed by #5412 but after testing updating to 10.2.0, the same error described in the original issue still occurs.

**Comment by DivyamAhuja** [comment]
```PS
  src\math\p5.Vector.js
    toString documentation
p5.Vector Object : [20, 30, 0]
      ✓ example #1 works
      ✓ example #2 works
    set documentation
      ✓ example #1 works
      ✓ example #2 works
    copy documentation
true
      ✓ example #1 works
    add documentation
      ✓ example #1 works
      ✓ example #2 works
Fatal error: Protocol error (Runtime.callFunctionOn): Object reference chain is too long
```

am still getting this error

**Comment by limzykenneth** [comment]
@DivyamAhuja Yes, that's what was mentioned in terms of the original error. Reverting to node 14 should work. We're looking into solutions for node 16.

**Comment by limzykenneth** [comment]
Good news and bad news. Good news: I found the source of this problem. Bad news: may be a bit tricky to resolve.

For a more detailed explanation see the issue on puppeteer: https://github.com/puppeteer/puppeteer/issues/7570. Essentially the problem occurs because there's a cyclic reference in p5.Vector instances, specifically it came from the `p5` property in the object itself (which is a reference to the sketch instance and nested in there is a cyclic reference). The cyclic reference causes a stringify operation to fail in puppeteer which in node 14 and before throws a "Unhandled promise rejection" message but carries on, from node 15 onwards an unhandled promise rejection is a fatal error which stops everything, thus what we see here.

There are a few ways that we can possibly resolve this. First way is to not have the cyclic reference in the first place (it came from `p5._curElement` which has a reference to itself) which I think would be best but might not be possible. Second  way is to not save the `p5` instance in `p5.Vector` instances, looking at the code the instance property is used to know how we are running the different functions in `p5.Vector` and to call two hidden/private functions `_toRadians` and `_fromRadians`. As such I think it is possible to remove the full reference to the `p5` instance from `p5.Vector` entirely without problems.

**Comment by limzykenneth** [comment]
TLDR: I manage to remove the reference to the p5 instance from p5.Vector objects and everything seems to work now. Will file a PR.

---

For posterity:

So I've looked into some potential solutions and the first is to remove cyclic reference entirely. This proves to be very difficult as not just `p5._curElement` has cyclic reference, `p5._elements`, `p5._renderer`, and `p5._preloadMethods` all have cyclic references as well. The one in `p5._curElement` is easy to remove but the rest, from what I can see, will require at least major changes to many core parts of p5, if such a thing is even possible without breaking things.

Another potential solution is to remove the reference to the p5 instance in the `p5.Vector` objects. These instance is used so that when `angleMode` is changed, methods such as `heading()` can return angles in the specified angle mode. This is done by directly calling the instance's `_fromRadians` and `_toRadians` method. As such there isn't really a need for `p5.Vector` to have a reference to the whole p5 instance. I've removed that reference and change the internal signature of the `p5.Vector` constructor to directly pass bounded versions of `_fromRadians` and `_toRadians` to the constructor. With this all tests passed on both node 14 and 16.
